src/coprocessor/dag/batch_executor/table_scan_executor.rs

```diff
@@ -0,0 +1,1107 @@
+                columns.push(LazyBatchColumn::raw_with_capacity(expect_rows));
+            }
+        } else {
+            // PK is unspecified in schema. All column should be in raw format.
+            for _ in 0..columns_len {
+                columns.push(LazyBatchColumn::raw_with_capacity(expect_rows));
+            }
+        }
+
+        LazyBatchColumnVec::from(columns)
+    }
+
+    fn process_kv_pair(
+        &mut self,
+        key: &[u8],
+        value: &[u8],
+        columns: &mut LazyBatchColumnVec,
+    ) -> Result<()> {
+        use crate::coprocessor::codec::{datum, table};
+        use crate::util::codec::number;
+
+        let columns_len = self.schema.len();
+        let mut decoded_columns = 0;
+
+        if let Some(handle_index) = self.handle_index {
+            let handle_id = table::decode_handle(key)?;
+            // TODO: We should avoid calling `push_int` repeatly. Instead we should specialize
+            // a `&mut Vec` first. However it is hard to program due to lifetime restriction.
+            columns[handle_index]
+                .mut_decoded()
+                .push_int(Some(handle_id));
+            decoded_columns += 1;
+            self.is_column_filled[handle_index] = true;
+        }
+
+        if value.is_empty() || (value.len() == 1 && value[0] == datum::NIL_FLAG) {
+            // Do nothing
+        } else {
+            // The layout of value is: [col_id_1, value_1, col_id_2, value_2, ...]
+            // where each element is datum encoded.
+            // The column id datum must be in var i64 type.
+            let mut remaining = value;
+            while !remaining.is_empty() && decoded_columns < columns_len {
+                if remaining[0] != datum::VAR_INT_FLAG {
+                    return Err(box_err!("Unable to decode row: column id must be VAR_INT"));
+                }
+                remaining = &remaining[1..];
+                let column_id = box_try!(number::decode_var_i64(&mut remaining));
+                let (val, new_remaining) = datum::split_datum(remaining, false)?;
+                // Note: The produced columns may be not in the same length if there is error due
+                // to corrupted data. It will be handled in `ScanExecutor`.
+                let some_index = self.column_id_index.get(&column_id);
+                if let Some(index) = some_index {
+                    let index = *index;
+                    if !self.is_column_filled[index] {
+                        columns[index].push_raw(val);
+                        decoded_columns += 1;
+                        self.is_column_filled[index] = true;
+                    } else {
+                        // This indicates that there are duplicated elements in the row, which is
+                        // unexpected. We won't abort the request or overwrite the previous element,
+                        // but will output a log anyway.
+                        warn!(
```

Comment:
Do we need to use `error` instead of `warn`?

Reply:
Maybe? I choose warn casually.

Reply:
Since it's an unexcepted behavior.

Reply:
But we have an internal document said that error logs should be handled immediately by users? This error does not need users to take actions.